### ISSUE TITLE
README: What operator is actually used for matrix multiplication?

### ISSUE THREAD

**Issue by davidfstr** [anuranbaka/cppnumpy#22]
The README says:

> The operator "^" is used for matrix multiplication rather than bitwise XOR

But "^" _is_ the bitwise XOR operater... Perhaps you're actually using some other operator and "^" is a misprint?

Perhaps you are actually using the ["@" operator in Python](https://www.python.org/dev/peps/pep-0465/), which is reserved for matrix multiplication?

**Comment by davidfstr** [comment]
Oh I think I get what you were saying now. Apparently it's too early in the morning for me to read straight :) What I initially read was:

> The operator "^" is used for matrix multiplication rather than ^

...which didn't make any sense.

The following slight editorial change might be clearer:

> The operator "^" is used for matrix multiplication rather than 👉performing👈 bitwise XOR

**Comment by anuranbaka** [comment]
How about
The operator "^" is overloaded to perform matrix multiplication rather than
bitwise XOR.
I feel like calling it an overload should get the reader in the right
mindset before they hit the end of the sentence.
We could also consider the longer form
The operator "^" is overloaded to perform matrix multiplication rather than
bitwise XOR, because C++ doesn't support the @ operator, and matrix
multiply is far more common than the rarely used XOR.
but I think the latter is too verbose, and the reasoning should be
basically obvious - that we stole a nearly useless operator to simplify a
common operation (like how streams stole "<<". We considered doing that
too.)

**Comment by davidfstr** [comment]
> How about - The operator "^" is overloaded to perform matrix multiplication rather - than bitwise XOR. I feel like calling it an overload should get the reader in the right mindset before they hit the end of the sentence.

👍 

> because C++ doesn't support the @ operator, and matrix multiply is far more common than the rarely used XOR.

I don't think this is necessarily obvious: You're consciously breaking compatibility with NumPy which is a big deal. So this kind of reasoning might be worth adding as a footnote (via an asterisk, since GitHub Flavored Markdown doesn't support direct footnotes, the last time I checked). Or maybe elsewhere deeper in the documentation rather than in the introduction.

**Comment by davidfstr** [comment]
Related issue: Another thing I misread this morning until just now (which probably puts my last comment in more context) is that this is actually a **C++** library rather than a Python library.

In the README I scanned this:
> This python-compatible matrix header library was built using a similar structure to the NumPy array. 

but read:
> This python matrix header library was built using a similar structure to the NumPy array.

Oops. Maybe:

> This C++ matrix header library was built to be compatible with NumPy and uses a similar structure to the NumPy array.

Perhaps with an explanation of why that would be useful:

> ...which makes it easy to write C++ code from existing NumPy code, or by authors who already have experience with NumPy.

(Not sure if the above is accurate or not.)

If you can get sleepy/drunk me to read the write thing with a quick scan, you'll attract more distracted researchers from the internet. ;)

**Comment by davidfstr** [comment]
You might argue that the title of the library makes it obvious that this is a C++ library, but that's only if you parse the title as "cpp numpy" rather than "cppnum.py" which is exactly what I did here, as a more Python-focused dev (who often sees Python library that end in "py"). :)

**Comment by anuranbaka** [comment]
Ohhhh. Some very interesting insight there.
Also interesting that the intent of the library is clearly not conveyed. You're the first person we've sent it to without previously explaining that, so perhaps we should do that more often.

The motivation for the library is this: I wanted a way to bind C++ and Numpy code with that just worked, with no surprises or overhead. Numpy is the de-facto language of data science because it really simplifies development, but we often also need the performance of C++, so most serious users wind up writing C++ extensions in Eigen or other libraries and binding them to Numpy, which leads to mysterious and hard to find bugs because the conversion process is imperfect, with differing memory layout, reference counting/GC, alignment restrictions, etc. Our eventual goal with CppNumpy is to provide a standalone C++ matrix library that is byte compatible with Numpy, and has bindings that correctly propagate memory management and error systems between C++ and Python, so that developers can write code without thinking about where a matrix originated. Our library compiles separately from the bindings, so that developers can write code compatible with Numpy without pulling in dependencies. 

We consider it an explicit design goal that our library maps 1:1 to a basic subset of the Numpy API, and deviations from that are bugs. In the rare case that this isn't possible due to language constraints (such as the "@" operator) we aim for a close equivalent to minimize mental load when porting code between the two.

I wonder if you have suggestions on how we might convey this in the readme without over burdening the reader?

**Comment by davidfstr** [comment]
> bind C++ and Numpy code

Several times you use the term "bind". What does this mean? Have Numpy code that can be embedded in (i.e. called by) C++? Have C++ code that can be embedded in (i.e. called by) Numpy code? Both?

> a standalone C++ matrix library that is byte compatible with Numpy, and has bindings that correctly propagate memory management and error systems between C++ and Python, so that developers can write code without thinking about where a matrix originated.

That sounds like a pretty good tagline, pitch, or introductory paragraph right there. :)

> our library maps 1:1 to a basic subset of the Numpy API, and deviations from that are bugs. In the rare case that this isn't possible due to language constraints (such as the "@" operator) we aim for a close equivalent to minimize mental load when porting code between the two.

There's your second intro paragraph. :)

**Comment by anuranbaka** [comment]
Thanks!  
The confusion about the term "bind" is very informative to me. I wonder if there is a clearer way to express what we mean.

By bindings I meant that any code you write in C++ that takes in our matrices can be compiled with a special header we provide, and the resulting shared library will be importable as a module in CPython, and the functions/classes will take and return Numpy arrays wherever our matrix class was used. The header uses C++11 template magic to make an implicit conversion to and from PyObject*, as well as linking up the reference counters and error propagation systems. In theory, we can also launch the CPython interpreter and run code from C++, but I haven't ever tried that.

**Comment by davidfstr** [comment]
> any code you write in C++ that takes in our matrices can be compiled with a special header we provide, and the resulting shared library will be importable as a module in CPython, and the functions/classes will take and return Numpy arrays wherever our matrix class was used.

There's another candidate for a 2nd or 3rd intro paragraph. :)
